openerp/osv/fields.py

```diff
@@ -861,8 +861,13 @@ def set(self, cr, obj, id, field, values, user=None, context=None):
                     # Must use write() to recompute parent_store structure if needed
                     obj.write(cr, user, act[2], {self._fields_id:id}, context=context or {})
                     ids2 = act[2] or [0]
-                    cr.execute('select id from '+_table+' where '+self._fields_id+'=%s and id <> ALL (%s)', (id,ids2))
-                    ids3 = map(lambda x:x[0], cr.fetchall())
+                    # if the o2m has a static domain we must respect it when unlinking
+                    domain = (self._domain(original_obj)
+                              if callable(self._domain) else self._domain)
+                    extra_domain = domain or []
+                    ids3 = obj.search(cr, user, [(self._fields_id,'=',id), ('id','not in',ids2)] + extra_domain, context=context)
```

Comment:
Is this going to affect performance in case no domain is defined?

Reply:
I'm not sure I understand your concern. The domain will be at least `[(self._fields_id,'=',id), ('id','not in',ids2)]`

Reply:
Previously, the query was done directly in SQL, and now is done in ORM. The question is if this affects something in the performance. If it does, maybe we should keep the query in SQL when no domain is found, and only switch to ORM when there's a domain.

Reply:
Yep, this was a concern. We will test this internally for a couple of days/weeks before merging anything anyway.

Reply:
OK, thanks.